controlplane/internal/controller/imageregistry/image_registry.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	additionalRegistryConfigMap, err := createImageRegistryConfig(imageRegistryConfigMap, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if err := client.Create(ctx, additionalRegistryConfigMap); err != nil {
+		return "", err
+	}
+	return additionalRegistryConfigMap.Name, nil
+}
+
+func createImageRegistryConfig(imageRegistry *corev1.ConfigMap, namespace string) (*corev1.ConfigMap, error) {
+	imageConfig := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      imageConfigMapName,
+			Namespace: namespace,
+		},
+		Data: make(map[string]string),
+	}
+	registryConf, ok := imageRegistry.Data[registryConfKey]
+	if ok {
+		imageDigestMirrorSet, err := getImageRegistries(registryConf)
+		if err != nil {
+			return nil, err
+		}
+		imageConfig.Data[imageDigestMirrorSetKey] = imageDigestMirrorSet
+	}
+
+	registryCert, ok := imageRegistry.Data[registryCertKey]
+	if ok {
+		addImageRegistryCertificate(imageConfig, registryCert)
+	}
+	return imageConfig, nil
+}
+
+// getImageRegistries reads a toml tree string with the structure:
+//
+// [[registry]]
+//
+//	location = "source-registry"
+//
+//	 [[registry.mirror]]
+//	 location = "mirror-registry"
+//
+// and converts it to an ImageDigestMirrorSet CR
+// and returns it as a marshalled JSON string
+func getImageRegistries(registry string) (string, error) {
+	// Parse toml and add mirror registry to image digest mirror set
+	tomlTree, err := toml.Load(registry)
+	if err != nil {
+		return "", err
+	}
+
+	idmsMirrors := make([]configv1.ImageDigestMirrors, 0)
+	registriesTree, ok := tomlTree.Get("registry").([]*toml.Tree)
+	if !ok {
+		return "", fmt.Errorf("failed to find registry in toml tree")
+	}
+
+	// Add each registry and its mirrors as a image digest mirror object
+	for _, registryTree := range registriesTree {
+		var imageDigestMirror configv1.ImageDigestMirrors
+		source, ok := registryTree.Get("location").(string)
+		if !ok {
+			continue
+		}
+		imageDigestMirror.Source = source
+
+		var mirrors []configv1.ImageMirror
+		mirrorTrees, ok := registryTree.Get("mirror").([]*toml.Tree)
+		if !ok {
+			continue
+		}
+
+		for _, mirrorTree := range mirrorTrees {
+			mirror, ok := mirrorTree.Get("location").(string)
+			if ok {
+				mirrors = append(mirrors, configv1.ImageMirror(mirror))
+			}
+		}
+
+		if len(mirrors) < 1 {
+			continue
+		}
+		imageDigestMirror.Mirrors = mirrors
+		idmsMirrors = append(idmsMirrors, imageDigestMirror)
+	}
+
+	if len(idmsMirrors) < 1 {
+		return "", fmt.Errorf("failed to find any mirror registry")
+	}
+
+	idms := configv1.ImageDigestMirrorSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ImageDigestMirrorSet",
+			APIVersion: configv1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: imageRegistryName,
+		},
+		Spec: configv1.ImageDigestMirrorSetSpec{
+			ImageDigestMirrors: idmsMirrors,
+		},
+	}
+
+	idmsJSON, err := json.Marshal(idms)
+	if err != nil {
+		return "", err
+	}
+
+	return string(idmsJSON), nil
+}
```

Comment:
I think this can be broken down into multiple smaller functions.

I see:
* getImageRegistries
* one (funcA) that wraps the logic returning `idmsMirrors`
* another one within funcA (funcB) that could return `imageDigestMirror` object

Reply:
Modified in this PR https://github.com/openshift-assisted/cluster-api-agent/pull/24